README.md
# CSY-DICOMweb-QIDO-RS
This repository generates the URL and requests PACS for DICOMweb QIDO-RS.
This repository was built by typescript.

## DICOM QIDO Parameter
### Required Matching Attributes
> https://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_10.6.html
DICOM PS3.18 2022b => 10.6 Search Transaction => Table 10.6.1-5. Required Matching Attributes
> https://dicom.nema.org/medical/dicom/2022b/output/chtml/part18/sect_10.6.html
DICOM PS3.18 2022b => 10.6 Search Transaction => Table 10.6.1-5. Required Matching Attributes

### Query Parameter Syntax (Terminology)
> https://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_8.3.4.html#sect_8.3.4.1
DICOM PS3.18 2022b => 8.3 Query Parameters => Table 8.3.4-1. Query Parameter Syntax
> https://dicom.nema.org/medical/dicom/2022b/output/chtml/part18/sect_8.3.4.html#sect_8.3.4.1
DICOM PS3.18 2022b => 8.3 Query Parameters => Table 8.3.4-1. Query Parameter Syntax

# How to use : Typescript
```typescript
// 引入套件
import QIDO from "csy-dicomweb-qido-rs";
import _ from "lodash";


// 實體化
const qido = new QIDO();

myFun();

async function myFun() {

    // "必須"自己初始化
    await qido.init();

    // 查詢階層設定：studies、series、instances
    qido.queryLevel = "studies";

    // 有使用到的套件參數設定：url-parse package
    qido.hostname = "test.dicom.tw";
    qido.pathname = "/dicomWeb";
    qido.protocol = "http";
    qido.port = "999";


    // 查詢參數設定：DICOM QIDO-RS Parameter
    let tempQueryParameter = {};

    tempQueryParameter = _.set(tempQueryParameter, "PatientID", "*");
    tempQueryParameter = _.set(tempQueryParameter, "limit", "10");
    tempQueryParameter = _.set(tempQueryParameter, "offset", "0");

    // 查詢參數用物件套入
    qido.queryParameter = tempQueryParameter;


    // 設定 Token:現在尚未啟用
    // let myHeaders = {};
    // myHeaders = _.set(myHeaders, "token", "jf903j2vunf9843nvyf934qc");
    // await qido.setUseToken(myHeaders);


    //查詢 同步模式
    await qido.query();
    console.log(qido.response);
}

```


# How to use : Javascript
```javascript 
import QIDO from "csy-dicomweb-qido-rs";

(async () => {
//實體化
let qido = new QIDO();

//"必須"自己初始化
await qido.init();

//查詢階層設定：studies、series、instances
qido.queryLevel = "studies";

//有使用到的套件參數設定：url-parse package
qido.hostname = "test.dicom.tw";
qido.pathname = "/dicomWeb";
qido.protocol = "http";
qido.port = "999";

//查詢參數設定：DICOM QIDO-RS Parameter
let tempQueryParameter = {};
tempQueryParameter.PatientID = '*';
tempQueryParameter.limit = "10";
tempQueryParameter.offset = "0";

//查詢參數用物件套入
qido.queryParameter = tempQueryParameter;

//設定 Token:現在尚未啟用
// let myHeaders = {};

// myHeaders.token = "jf903j2vunf9843nvyf934qc";
// await qido.setUseToken(myHeaders);

//查詢 同步模式
await qido.query();

//印出 response: json
console.log(qido.response);

})()
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.3](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.3.2...v0.3.3) (2022-07-29)

### [0.3.2](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.3.1...v0.3.2) (2022-07-29)

### [0.3.1](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.3.0...v0.3.1) (2022-07-29)

## [0.3.0](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.8...v0.3.0) (2022-07-29)


### ⚠ BREAKING CHANGES

* building with typescript.

* change to generate by typescript ([00b937a](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/commit/00b937aea40de92e491603620578cf39c3ee30f9))

### [0.2.8](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.7...v0.2.8) (2022-07-13)

### [0.2.7](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.6...v0.2.7) (2022-07-13)

### [0.2.6](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.5...v0.2.6) (2022-07-13)

### [0.2.5](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.4...v0.2.5) (2022-07-13)

### [0.2.4](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.3...v0.2.4) (2022-07-13)

### [0.2.3](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.2...v0.2.3) (2022-07-13)

### [0.2.2](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.1...v0.2.2) (2022-07-12)

### [0.2.1](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.2.0...v0.2.1) (2022-07-12)


### Bug Fixes

* use setUseToken will alret user ([2ba916f](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/commit/2ba916fdffa3d07e9fac9ea1927ac25f2773a683))

## [0.2.0](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.1.5...v0.2.0) (2022-07-12)


### Features

* studies level QIDO-RS retrieval is enable ([86b4adb](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/commit/86b4adb9d71026521cb113523766058b9a2c630f))

### [0.1.5](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.1.4...v0.1.5) (2022-07-12)

### [0.1.4](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.1.3...v0.1.4) (2022-07-11)


### Features

* change package.json main property to DICOMwebQIDORS.js from index.js ([4cac0c7](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/commit/4cac0c765a6cbf2941f6f632f96c1fbb307267ee))

### [0.1.3](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.1.2...v0.1.3) (2022-07-11)


### Features

* fix url-parse package to dependncies from devDependenies ([63557e7](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/commit/63557e7d9dc3170cefc8164a574758a1355da39a))

### [0.1.2](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/compare/v0.1.1...v0.1.2) (2022-07-11)

### 0.1.1 (2022-07-11)


### Features

* add commitlint & husky to generate conventional commit message ([70b7fa9](https://github.com/JieuTang/CSY-DICOMweb-QIDO-RS/commit/70b7fa9e7ee23701a9b412618bcc9526c04f9a00))


test.js
import QIDO from "./lib/DICOMwebQIDORS.js";

(async () => {
//實體化
let qido = new QIDO();

//"必須"自己初始化
await qido.init();

//查詢階層設定：studies、series、instances
qido.queryLevel = "studies";

//有使用到的套件參數設定：url-parse package
qido.hostname = "hackathon.raccoon.dicom.tw";
qido.pathname = "/dicom-web";
qido.protocol = "https";
qido.port = "443";

//查詢參數設定：DICOM QIDO-RS Parameter
let tempQueryParameter = {};
tempQueryParameter.PatientID = '*';
tempQueryParameter.limit = "10";
tempQueryParameter.offset = "0";

//查詢參數用物件套入
qido.queryParameter = tempQueryParameter;

//設定 Token:現在尚未啟用
// let myHeaders = {};

// myHeaders.token = "jf903j2vunf9843nvyf934qc";
// await qido.setUseToken(myHeaders);

//查詢 同步模式
await qido.query();

//印出 response: json
console.log(qido.response);
})()

src/DICOMwebQIDORS.ts
import URL from "url-parse";
import _ from "lodash";
import axios, { AxiosRequestHeaders } from "axios";

import QIDOParameter from "./config/QIDOParameter.config";
import QueryLevel from "./config/QueryLevel.config";
import AllowProtocol from "./config/AllowProtocol.config";

class DICOMwebQIDORS {

    // QIDOParameter.json
    terminologyList: object = {};
    studyParameterList: object = {};
    seriesParameterList: object = {};
    instanceParameterList: object = {};

    // Token
    isUseToken: boolean = false;
    tokenObject: object = {};

    // url-parse package
    hostname: string = "";
    pathname: string = "";
    port: string = "";
    protocol: string = "";
    queryParameter: object = {}

    // Query Level
    queryLevel: string = "";

    // Response
    response: object = {};


    async init() {
        await this._loadQIDOParameter();
    }

    async query() {
        await this._validateQueryLevel();
        await this._validateUrlComponent();

        if (this.queryLevel === QueryLevel.studies) {
            await this._queryStudies();
        } else if (this.queryLevel === QueryLevel.series) {
            await this._querySeries();
        } else if (this.queryLevel === QueryLevel.instances) {
            await this._queryInstances();
        } else {
            console.log("queryLevel Error");
        }
    }

    async setUseToken(tokenObject: object) {

        // tokenObject 不是 Object 就跳錯誤
        if (!(_.isObject(tokenObject))) {
            console.log("tokenValue must be object type.");
        }

        this.isUseToken = true;
        this.tokenObject = tokenObject;
        console.log("Here is function setUseToken(). This function is not enable in this version.");
    }



    async _loadQIDOParameter() {
        const _terminoloy = _.get(QIDOParameter, "Terminology");
        const _study = _.get(QIDOParameter, "Study");
        const _series = _.get(QIDOParameter, "Series");
        const _instance = _.get(QIDOParameter, "Instance");

        this.terminologyList = _.cloneDeep(_terminoloy);
        this.studyParameterList = _.cloneDeep(await this._getCombinedObjectWithInvertObject(_study));
        this.seriesParameterList = _.cloneDeep(await this._getCombinedObjectWithInvertObject(_series));
        this.instanceParameterList = _.cloneDeep(await this._getCombinedObjectWithInvertObject(_instance));
    }


    async _getCombinedObjectWithInvertObject(object1: object) {
        return _.assign(object1, _.invert(object1));
    }


    async _validateQueryLevel() {
        // falsy value: null、undefined、NaN、emptyString、0、false
        // https://262.ecma-international.org/5.1/#sec-9.2
        // 如果 queryLevel 數值是 falsy 值，就代表是錯的。
        if (!(_.get(QueryLevel, this.queryLevel))) {
            console.log("QueryLevel Value Error! \nQueryLevel ValueSet is [" + _.keys(QueryLevel) + "]");
        }
    }

    async _validateUrlComponent() {
        // falsy value: null、undefined、NaN、emptyString、0、false
        // https://262.ecma-international.org/5.1/#sec-9.2
        // 如果 hostname、pathname、protocol 數值是 falsy 值，就代表是錯的。
        if (!(this.hostname && this.pathname && this.protocol)) {
            console.log("hostname、pathname、protocol, three value can not be falsy.\n falsy value: null、undefined、NaN、emptyString、0、false");
        }

        // protocol 必須在清單內。
        if (!(_.get(AllowProtocol, this.protocol))) {
            console.log("Protocol Error.\nProtocol ValueSet is [" + _.keys(AllowProtocol) + "]");
        }

        // port 必須是字串型態的數字
        // 不是字串 && 轉成數字後不是整數，跳出錯誤。
        if (!(_.isString(this.port) && (_.isInteger((_.toNumber(this.port)))))) {
            console.log("port value has be String type and Integer value. Example '443' or '80'. ");
        }

        // queryParameter 必須在字典內
        const legalKeySet = _.keys(_.assign(this.terminologyList, this.studyParameterList, this.seriesParameterList, this.instanceParameterList));
        _.forEach(this.queryParameter, (value, key) => {
            if (!(_.includes(legalKeySet, key))) {
                console.log("Key value: " + _.toString(key) + " is no allow.");
            }
        })

    }




    async _queryStudies() {
        const url = new URL("");
        url.set('hostname', this.hostname);
        url.set('pathname', this.pathname + "/" + QueryLevel.studies);
        url.set('port', this.port);
        url.set('protocol', _.get(AllowProtocol, this.protocol));
        url.set('query', this.queryParameter);

        console.log(url.toString());
        this.response = await this._getRequestResponse(url.toString());
    }

    async _querySeries() {
        console.log("Here is Query Series Level in QIDO-RS. This function is not enable in this version.");
    }

    async _queryInstances() {
        console.log("Here is Query Instances Level in QIDO-RS. This function is not enable in this version.");
    }

    async _getRequestResponse(inputUrl: string): Promise<object> {
        let result: object = {};
        let response = {};

        if (this.isUseToken) {
            await axios({
                method: "get",
                url: inputUrl,
                headers: this.tokenObject as AxiosRequestHeaders
            }).then(
                (res) => {
                    response = _.cloneDeep(res.data);
                })
                .catch((error) => console.log(error))

        } else {
            await axios({
                method: "get",
                url: inputUrl,
            }).then(
                (res) => {
                    response = _.cloneDeep(res.data);
                })
                .catch((error) => console.log(error))



            // try {
            //     response = await (await axios.get(url)).data;
            //     result = response;
            // } catch (error) {
            //     if (axios.isAxiosError(error)) {
            //         console.log(error);
            //     }
            // }
        }

        result = _.cloneDeep(response);

        return result;
    }
}

export default DICOMwebQIDORS;

src/config/QIDOParameter.config.ts
export default {
    "Terminology": {
        "match": "match",
        "fuzzymatching": "fuzzymatching",
        "includefield": "includefield",
        "limit": "limit",
        "offset": "offset"
    },
    "Study": {
        "StudyDate": "00080020",
        "StudyTime": "00080030",
        "AccessionNumber": "00080050",
        "ModalitiesInStudy": "00080061",
        "ReferringPhysicianName": "00080090",
        "PatientName": "00100010",
        "PatientID": "00100020",
        "StudyInstanceUID": "0020000D",
        "StudyID": "00200010"
    },
    "Series": {
        "Modality": "00080060",
        "SeriesInstanceUID": "0020000E",
        "SeriesNumber": "00200011",
        "PerformedProcedureStepStartDate": "00400244",
        "PerformedProcedureStepStartTime": "00400245",
        "RequestAttributesSequence": "00400275",
        "ScheduledProcedureStepID": "00400009",
        "RequestedProcedureID": "00401001"
    },
    "Instance": {
        "SOPClassUID": "00080016",
        "SOPInstanceUID": "00080018",
        "InstanceNumber": "00200013"
    }
}



DICOMwebQIDORS.js
import URL from "url-parse";
import _ from "lodash";
import fetch from 'node-fetch';

import QIDOParameter from "./Config/QIDOParameter.json" assert {type: "json"};
import QueryLevel from "./Config/QueryLevel.json" assert {type: "json"};
import AllowProtocol from "./Config/AllowProtocol.json" assert {type: "json"};

class DICOMwebQIDORS {
    constructor() {
        //QIDOParameter.json
        this.terminologyList = undefined;
        this.studyParameterList = undefined;
        this.seriesParameterList = undefined;
        this.instanceParameterList = undefined;
        
        //Token
        this.isUseToken = false;
        this.tokenObject = undefined;

        //url-parse package
        this.hostname = undefined;
        this.pathname = undefined;
        this.port = undefined;
        this.protocol = undefined;
        this.queryParameter = undefined

        //Query Level
        this.queryLevel = undefined;

        //Response
        this.response = undefined;
    }

    async init() {
        await this._loadQIDOParameter();
    }

    async query() {
        await this._validateQueryLevel();
        await this._validateUrlComponent();

        if (this.queryLevel === QueryLevel.studies) {
            await this._queryStudies();
        } else if (this.queryLevel === QueryLevel.series) {
            await this._querySeries();
        } else if (this.queryLevel === QueryLevel.instances) {
            await this._queryInstances();
        } else {
            throw "queryLevel Error";
        }
    }

    async setUseToken(tokenObject) {

        //tokenObject 不是 Object 就跳錯誤
        if(!(_.isObject(tokenObject))) {
            throw "tokenValue must be object type."
        }

        this.isUseToken = true;
        this.tokenObject = tokenObject;
        throw "Here is function setUseToken(). This function is not enable in this version.";
    }

    async _loadQIDOParameter() {
        let _terminoloy = _.get(QIDOParameter, "Terminology");
        let _study = _.get(QIDOParameter, "Study");
        let _series = _.get(QIDOParameter, "Series");
        let _instance = _.get(QIDOParameter, "Instance");

        this.terminologyList = _.cloneDeep(_terminoloy);
        this.studyParameterList = _.cloneDeep(await this._getCombinedObjectWithInvertObject(_study));
        this.seriesParameterList = _.cloneDeep(await this._getCombinedObjectWithInvertObject(_series));
        this.instanceParameterList = _.cloneDeep(await this._getCombinedObjectWithInvertObject(_instance));
    }

    async _getCombinedObjectWithInvertObject(object1) {
        return _.assign(object1, _.invert(object1));
    }

    async _validateUrlComponent() {
        //falsy value: null、undefined、NaN、emptyString、0、false
        //https://262.ecma-international.org/5.1/#sec-9.2
        //如果 hostname、pathname、protocol 數值是 falsy 值，就代表是錯的。
        if (!(this.hostname && this.pathname && this.protocol)) {
            throw "hostname、pathname、protocol, three value can not be falsy.\n falsy value: null、undefined、NaN、emptyString、0、false";
        }

        //protocol 必須在清單內。
        if (!(_.get(AllowProtocol, this.protocol))) {
            throw "Protocol Error.\nProtocol ValueSet is [" + _.keys(AllowProtocol) + "]";
        }

        //port 必須是字串型態的數字
        //不是字串 && 轉成數字後不是整數，跳出錯誤。
        if (!(_.isString(this.port) && (_.isInteger((_.toNumber(this.port)))))) {
            throw "port value has be String type and Integer value. Example '443' or '80'. ";
        }

        //queryParameter 必須在字典內
        let legalKeySet = _.keys(_.assign(this.terminologyList, this.studyParameterList, this.seriesParameterList, this.instanceParameterList));
        _.forEach(this.queryParameter, (value, key) => {
            if (!(_.includes(legalKeySet, key))) {
                throw "Key value: " + _.toString(key) + " is no allow.";
            }
        })
        
    }

    async _validateQueryLevel() {
        //falsy value: null、undefined、NaN、emptyString、0、false
        //https://262.ecma-international.org/5.1/#sec-9.2
        //如果 queryLevel 數值是 falsy 值，就代表是錯的。
        if (!(_.get(QueryLevel, this.queryLevel))) {
            throw "QueryLevel Value Error! \nQueryLevel ValueSet is [" + _.keys(QueryLevel) + "]";
        }
    }

    async _queryStudies() {
        let url = new URL();
        url.set('hostname', this.hostname);
        url.set('pathname', this.pathname + "/" + QueryLevel.studies);
        url.set('port', this.port);
        url.set('protocol', AllowProtocol[this.protocol]);
        url.set('query', this.queryParameter);

        console.log(url.toString());
        this.response = await this._getRequestResponse(url.toString());
    }

    async _querySeries() {
        throw "Here is Query Series Level in QIDO-RS. This function is not enable in this version.";
    }

    async _queryInstances() {
        throw "Here is Query Instances Level in QIDO-RS. This function is not enable in this version.";
    }

    async _getRequestResponse(url) {
        let result = undefined;
        let response = undefined;
        
        if (this.isUseToken) {
            response = await fetch(url, {
                headers: this.tokenObject
            });
        } else {
            response = await fetch(url);
        }

        result = await response.json();

        return result;
    }
}

export default DICOMwebQIDORS;